### PR TITLE
feat(forms): allow Textarea to dynamically change height

### PR DESCRIPTION
**Pull request by austin94**
## Description

This PR introduces the concept of auto-resizing to the `Textarea` component. This functionality is disabled by default.

### Props

This feature is enabled with two new props, `minRows` and `maxRows`. These new props define the upper and lower bounds for overflow when resizing and only enable auto-resizing if the `isResizable` props is not provided.

### Size Calculations

After looking at several other implementations, the most common way of achieving this functionality in a cross-browser compatible way is to use hidden "shadow" input with the same styling to calculate when an overflow occurs. This shadow input is only rendered if the auto-resizing props are provided.

Resizing is handled on initial render, when the screen is resized (with debounce), and when the input value is changed. It also handles both controlled and uncontrolled usages.

The use of the shadow input has some great results in mobile browsers when compared against simpler size calculations.

### Testing

JSDOM makes it difficult to mock scrollHeight and other computed style properties. I was able to test basic functionality, but some of the more complex edge-cases are only reproducible in a browser environment.

## Checklist

- [ ] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :metal: renders as expected with Bedrock CSS (`?bedrock`)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

### CODE REVIEW

File: packages/forms/src/elements/Textarea.tsx [hunk 1]
```diff
@@ -20,27 +22,160 @@ export interface ITextareaProps extends TextareaHTMLAttributes<HTMLTextAreaEleme
...
+          <StyledTextarea
+            aria-hidden
+            readOnly
+            isHidden
+            className={props.className}
+            ref={shadowTextAreaRef}
+            tabIndex={-1}
+            style={style}
+          />
+        ) : null}
```
Comment (jzempel) [inline packages/forms/src/elements/Textarea.tsx:174]: Is this more readable as `{isAutoResizable && <StyledTextarea... />}`?

Comment (austin94) [reply packages/forms/src/elements/Textarea.tsx:174]: I just stumbled onto https://kentcdodds.com/blog/use-ternaries-rather-than-and-and-in-jsx and it had some good points against `&&` in render functions.

I can update this if you feel strongly against it.

Comment (jzempel) [reply packages/forms/src/elements/Textarea.tsx:174]: The article is guarding against array checking. As this is a boolean, I'd like to think `&&` is a reliable choice.

Comment (austin94) [reply packages/forms/src/elements/Textarea.tsx:174]: Updated.

File: packages/forms/src/elements/Textarea.tsx [hunk 2]
```diff
@@ -20,27 +22,160 @@ export interface ITextareaProps extends TextareaHTMLAttributes<HTMLTextAreaEleme
...
+      if (!isAutoResizable) {
+        return;
+      }
+
+      const textarea = textAreaRef.current!;
+      const computedStyle = window.getComputedStyle(textarea);
+      const shadowTextArea = shadowTextAreaRef.current!;
+
+      shadowTextArea.style.width = computedStyle.width;
+      shadowTextArea.value = textarea.value || textarea.placeholder || ' ';
```
[the remark this answers is not in the document]
Comment (austin94) [reply packages/forms/src/elements/Textarea.tsx:63]: There was an issue relating to this where I wasn't passing `isBare` and `isCompact` to the shadow element. This is now resolved, but doesn't address this specific issue.

The textarea inherits it's min-height from the `StyledTextInput` component. This min-height is causing the shadow element to pin it's height when the font-size is smaller. If you increase the font-size the value is calculated correctly. When the `isBare` prop is provided the font-size is handled well.

Would we want to remove the `min-height` value for the `Textarea` to support this use-case?

[the remark this answers is not in the document]
Comment (jzempel) [reply packages/forms/src/elements/Textarea.tsx:63]: No, removing `min-height` is not desirable. I think it's ok to get close on this one, with the expectation that folks won't typically be manipulating the `font-size`.
